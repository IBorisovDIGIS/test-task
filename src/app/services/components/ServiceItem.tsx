import React, { useState } from 'react';
import {
  Button, Grid, InputAdornment, makeStyles, Paper, TextField, Tooltip, Typography,
} from '@material-ui/core';
import copy from 'clipboard-copy';
import clsx from 'clsx';
import { useTranslation } from 'react-i18next';
import { black, blue, white } from '../../../constants/colors';
import CustomIcon from '../../common/CustomIcon';
import CopyIcon from '../../../assets/icons/copy.svg';
import { IBounce } from '../../../interfaces';

const useStyles = makeStyles((theme) => ({
  container: {
    padding: '26px 32px',
    marginTop: 24,
  },
  containerActivated: {
    backgroundColor: black,
  },
  button: {
    padding: '8px 10px',
  },
  title: {
    fontWeight: 300,
    fontSize: 29,
    color: '#000',
  },
  subtitle: {
    fontSize: 15,
    color: '#7D7D7D',
  },
  textActivated: {
    color: white,
  },
  bonusButton: {
    height: 48,
    backgroundColor: blue,
    color: white,
  },
  bonusButtonActivated: {
    backgroundColor: white,
    color: blue,
  },
  input: {
    borderRadius: 4,
    backgroundColor: theme.palette.common.white,
    border: '1px solid #ced4da',
    fontSize: 16,
    padding: '5px 10px',
  },
  copyIconWrapper: {
    height: '100%',
    width: 25,
  },
  copyIcon: {
    cursor: 'pointer',
  },
}));

const ServiceItem: React.FC<IBounce> = ({ description, promocode, title }:IBounce) => {
  const classes = useStyles();
  const { t } = useTranslation();

  const [isCopied, setIsCopied] = useState(false);
  const [isActivated, setIsActivated] = useState(false);

  const toggleBonus = () => setIsActivated((prevIsActivated) => !prevIsActivated);

  const handleOnCopy = async () => {
    await copy(promocode);
    setIsCopied(true);
    setTimeout(() => setIsCopied(false), 1000);
  };

  const renderCopyButton = () => (
    <Tooltip
      open={isCopied}
      title={promocode}
      placement="top"
      arrow
    >
      <InputAdornment
        position="end"
        onClick={handleOnCopy}
        className={classes.copyIconWrapper}
      >
        <CustomIcon
          icon={CopyIcon}
          iconClasses={classes.copyIcon}
        />
      </InputAdornment>
    </Tooltip>
  );

  return (
    <Grid item xs={12}>
      <Paper classes={{ root: clsx(classes.container, isActivated && classes.containerActivated) }}>
        <Grid container spacing={2} alignItems="stretch">
          <Grid item xs={6}>
            <Typography
              classes={{ root: clsx(classes.title, isActivated && classes.textActivated) }}
            >
              {title}
            </Typography>
            <Typography
              classes={{ root: clsx(classes.subtitle, isActivated && classes.textActivated) }}
            >
              {description}
            </Typography>
          </Grid>
          <Grid item xs={3} container alignItems="flex-end">
            <TextField
              fullWidth
              label={t('promocode').toUpperCase()}
              disabled
              value={promocode}
              InputProps={{
                disableUnderline: true,
                classes: { root: classes.input },
                endAdornment: renderCopyButton(),
              }}
            />
          </Grid>
          <Grid item xs={3} container alignItems="flex-end">
            <Button
              variant="contained"
              fullWidth
              classes={{
                root: clsx(classes.bonusButton, isActivated && classes.bonusButtonActivated),
              }}
              onClick={toggleBonus}
            >
              {isActivated ? t('deActivateBonus') : t('activateBonus')}
            </Button>
          </Grid>
        </Grid>
      </Paper>
    </Grid>
  );
};

export default ServiceItem;
