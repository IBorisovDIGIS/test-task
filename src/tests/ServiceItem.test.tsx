import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {
  Button, InputAdornment,
  TextField, Typography,
} from '@material-ui/core';
import toJson from 'enzyme-to-json';
import ServiceItem from "../app/services/components/ServiceItem";
import data from '../assets/data.json';

configure({ adapter: new Adapter() });

jest.mock('react-i18next', () => ({
  useTranslation: () => ({ t: (key: string) => key })
}));


describe('Search filter should render correctly', () => {
  const { bonuses } = data;
  const serviceItemData = bonuses[0];

  it('Should renders correctly', () => {
    const container = shallow(<ServiceItem {...serviceItemData} />);

    expect(toJson(container)).toMatchSnapshot();
  });

  it('Should render values correctly', () => {
    const container = shallow(<ServiceItem {...serviceItemData} />);
    const title = container.find(Typography).first().text()
    const description = container.find(Typography).at(1).text()
    const promocode = container.find(TextField).prop('value')
    expect(title).toEqual(serviceItemData.title);
    expect(description).toEqual(serviceItemData.description);
    expect(promocode).toEqual(serviceItemData.promocode);
  });

  it('Should change state by clicking on button', async () => {
    const container = shallow(<ServiceItem {...serviceItemData} />);

    expect(container.find(Button).text()).toEqual('activateBonus')
    container.find(Button).simulate('click');


    await new Promise(resolve =>  setTimeout(() => {
      expect(container.find(Button).text()).toEqual('deActivateBonus')
      resolve(true)
    }));
  });
});
