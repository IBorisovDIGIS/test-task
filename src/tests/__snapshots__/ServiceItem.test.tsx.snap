// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search filter should render correctly Should renders correctly 1`] = `
<WithStyles(ForwardRef(Grid))
  item={true}
  xs={12}
>
  <WithStyles(ForwardRef(Paper))
    classes={
      Object {
        "root": "makeStyles-container-1",
      }
    }
  >
    <WithStyles(ForwardRef(Grid))
      alignItems="stretch"
      container={true}
      spacing={2}
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
        xs={6}
      >
        <WithStyles(ForwardRef(Typography))
          classes={
            Object {
              "root": "makeStyles-title-4",
            }
          }
        >
          Sitecostructor.io
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(Typography))
          classes={
            Object {
              "root": "makeStyles-subtitle-5",
            }
          }
        >
          Site constructor
        </WithStyles(ForwardRef(Typography))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        alignItems="flex-end"
        container={true}
        item={true}
        xs={3}
      >
        <WithStyles(ForwardRef(TextField))
          InputProps={
            Object {
              "classes": Object {
                "root": "makeStyles-input-9",
              },
              "disableUnderline": true,
              "endAdornment": <WithStyles(ForwardRef(Tooltip))
                arrow={true}
                open={false}
                placement="top"
                title="itpaycodes"
              >
                <WithStyles(ForwardRef(InputAdornment))
                  className="makeStyles-copyIconWrapper-10"
                  onClick={[Function]}
                  position="end"
                >
                  <CustomIcon
                    alt=""
                    icon="copy.svg"
                    iconClasses="makeStyles-copyIcon-11"
                  />
                </WithStyles(ForwardRef(InputAdornment))>
              </WithStyles(ForwardRef(Tooltip))>,
            }
          }
          disabled={true}
          fullWidth={true}
          label="PROMOCODE"
          value="itpaycodes"
        />
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        alignItems="flex-end"
        container={true}
        item={true}
        xs={3}
      >
        <WithStyles(ForwardRef(Button))
          classes={
            Object {
              "root": "makeStyles-bonusButton-7",
            }
          }
          fullWidth={true}
          onClick={[Function]}
          variant="contained"
        >
          activateBonus
        </WithStyles(ForwardRef(Button))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Paper))>
</WithStyles(ForwardRef(Grid))>
`;
